pkg/utils/apply/apply.go

```diff
@@ -0,0 +1,205 @@
+package apply
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// An Object is a Kubernetes object.
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+// Applicator provides two approaches to applies changes to an object,
+// patching and updating.
+type Applicator interface {
+	PatchingApplicator
+	UpdatingApplicator
```

Comment:
Will we have scenario to use UpdatingApplicator? If not, I suggest we remove it.

Reply:
We can remove UpdatingApplicator. Even there's scenario where need updating-apply, someone can just use raw client.Update() directly.

Reply:
> someone can just use raw client.Update() directly.

Good idea!